<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>evolver Module &mdash; pyvolve 0.1 documentation</title>
    
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="pyvolve 0.1 documentation" href="index.html" />
    <link rel="up" title="Python modules in pyvolve" href="modules.html" />
    <link rel="prev" title="matrix_builder Module" href="matrix_builder.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="matrix_builder.html" title="matrix_builder Module"
             accesskey="P">previous</a> |</li>
        <li><a href="index.html">pyvolve 0.1 documentation</a> &raquo;</li>
          <li><a href="modules.html" accesskey="U">Python modules in <cite>pyvolve</cite></a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="module-evolver">
<span id="evolver-module"></span><h1><tt class="docutils literal"><span class="pre">evolver</span></tt> Module<a class="headerlink" href="#module-evolver" title="Permalink to this headline">¶</a></h1>
<p>The module will evolve sequences along a phylogeny.</p>
<dl class="class">
<dt id="evolver.Site">
<em class="property">class </em><tt class="descclassname">evolver.</tt><tt class="descname">Site</tt><a class="headerlink" href="#evolver.Site" title="Permalink to this definition">¶</a></dt>
<dd><p>Defines the Site class, which holds information for each evolved site.
Currently the sequence, but this will be expanded.</p>
</dd></dl>

<dl class="class">
<dt id="evolver.Evolver">
<em class="property">class </em><tt class="descclassname">evolver.</tt><tt class="descname">Evolver</tt><big>(</big><em>**kwargs</em><big>)</big><a class="headerlink" href="#evolver.Evolver" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <tt class="xref py py-class docutils literal"><span class="pre">object</span></tt></p>
<dl class="docutils">
<dt>This callable class evolves sequences along a phylogeny. By default, Evolver will evolve sequences and create several output files:</dt>
<dd><blockquote class="first">
<div><ol class="arabic">
<li><dl class="first docutils">
<dt>simulated_alignment.fasta</dt>
<dd><ul class="first last simple">
<li>The resulting simulated alignment</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>site_rates.txt </dt>
<dd><ul class="first last simple">
<li>File providing rate information about each simulated column. Gives the partition and rate cateogy for each site in final simulated alignment.</li>
<li>Tab-delimited file with fields, Site_Index    Partition_Index     Rate_Category . All indexing is from <em>1</em>.</li>
</ul>
</dd>
</dl>
</li>
<li><dl class="first docutils">
<dt>site_rates_info.txt</dt>
<dd><ul class="first last simple">
<li>File providing the true site-rate heterogeneity values (either the rate scaling factor or dN and dS) for each rate category.</li>
<li>Tab-delimited file with fields, Partition_Index    Model_Name    Rate_Category    Rate_Probability    Rate_Factor</li>
</ul>
</dd>
</dl>
</li>
</ol>
</div></blockquote>
<p class="last">Note that file creation may be suppressed or files may be renamed using optional arguments given below.</p>
</dd>
<dt>Required keyword arguments include,</dt>
<dd><ol class="first last arabic simple">
<li><strong>tree</strong> is the phylogeny (parsed with the <tt class="docutils literal"><span class="pre">newick.read_tree</span></tt> function) along which sequences are evolved</li>
<li><strong>partitions</strong> is a list of Partition instances to evolve</li>
</ol>
</dd>
<dt>Optional keyword arguments include,</dt>
<dd><ol class="first last arabic simple">
<li><strong>seqfile</strong> is a custom name for the output simulated alignment. Provide None or False to suppress file creation.</li>
<li><strong>seqfmt</strong>  is the format for seqfile (either fasta, nexus, phylip, phylip-relaxed, stockholm, etc. Anything that Biopython can accept!!) Default is FASTA.</li>
<li><strong>ratefile</strong> is a custom name for the &#8220;site_rates.txt&#8221; file. Provide None or False to suppress file creation.</li>
<li><strong>infofile</strong> is a custom name for the &#8220;site_rates_info.txt&#8221; file. Provide None or False to suppress file creation.</li>
<li><strong>write_anc</strong> is a boolean argument (True or False) for whether ancestral sequences should be output along with the tip sequences. Default is False.</li>
</ol>
</dd>
</dl>
<dl class="method">
<dt id="evolver.Evolver.__call__">
<tt class="descname">__call__</tt><big>(</big><big>)</big><a class="headerlink" href="#evolver.Evolver.__call__" title="Permalink to this definition">¶</a></dt>
<dd><p>Simulate sequences, perform any necessary post-processing, and save sequences and/or other info to appropriate files.</p>
<dl class="docutils">
<dt>Examples:</dt>
<dd><div class="first last highlight-python"><div class="highlight"><pre><span class="gp">&gt;&gt;&gt; </span><span class="c"># Evolve according to default settings</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">evolve</span> <span class="o">=</span> <span class="n">Evolver</span><span class="p">(</span><span class="n">tree</span> <span class="o">=</span> <span class="n">my_tree</span><span class="p">,</span> <span class="n">partitions</span> <span class="o">=</span> <span class="n">my_partition_list</span><span class="p">)()</span>

<span class="gp">&gt;&gt;&gt; </span><span class="c"># Include ancestral sequences in output file</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">evolve</span> <span class="o">=</span> <span class="n">Evolver</span><span class="p">(</span><span class="n">tree</span> <span class="o">=</span> <span class="n">my_tree</span><span class="p">,</span> <span class="n">partitions</span> <span class="o">=</span> <span class="n">my_partition_list</span><span class="p">,</span> <span class="n">write_anc</span> <span class="o">=</span> <span class="bp">True</span><span class="p">)()</span>

<span class="gp">&gt;&gt;&gt; </span><span class="c"># Custom sequence file name and format, and suppress rate information</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">evolve</span> <span class="o">=</span> <span class="n">Evolver</span><span class="p">(</span><span class="n">tree</span> <span class="o">=</span> <span class="n">my_tree</span><span class="p">,</span> <span class="n">partitions</span> <span class="o">=</span> <span class="n">my_partition_list</span><span class="p">,</span> <span class="n">seqfile</span> <span class="o">=</span> <span class="s">&quot;my_seqs.phy&quot;</span><span class="p">,</span> <span class="n">seqfmt</span> <span class="o">=</span> <span class="s">&quot;phylip&quot;</span><span class="p">,</span> <span class="n">ratefile</span> <span class="o">=</span> <span class="bp">None</span><span class="p">,</span> <span class="n">infofile</span> <span class="o">=</span> <span class="bp">None</span><span class="p">)()</span>
</pre></div>
</div>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h4>Previous topic</h4>
  <p class="topless"><a href="matrix_builder.html"
                        title="previous chapter"><tt class="docutils literal"><span class="pre">matrix_builder</span></tt> Module</a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/evolver.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="matrix_builder.html" title="matrix_builder Module"
             >previous</a> |</li>
        <li><a href="index.html">pyvolve 0.1 documentation</a> &raquo;</li>
          <li><a href="modules.html" >Python modules in <cite>pyvolve</cite></a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2015, Stephanie J. Spielman.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.3.
    </div>
  </body>
</html>